engine/access/rest/websockets/controller.go

```diff
@@ -193,20 +319,57 @@ func (c *Controller) handleListSubscriptions(ctx context.Context, msg models.Lis
 }
 
 func (c *Controller) shutdownConnection() {
-	defer close(c.communicationChannel)
-	defer func(conn *websocket.Conn) {
-		if err := c.conn.Close(); err != nil {
-			c.logger.Error().Err(err).Msg("error closing connection")
+	c.shutdownOnce.Do(func() {
+		c.shutdown = true
+
+		defer func() {
+			if err := c.conn.Close(); err != nil {
+				c.logger.Error().Err(err).Msg("error closing connection")
+			}
+			close(c.communicationChannel)
```

Comment:
it's not safe to close this here, because the data providrs could continue to write out new messages causing a panic. 

going back to the original design, we had said that data providers would get a callback function they could call to put new data on this queue. That would allow the controller to decide when to stop accepting new messages and to close this channel safely.

Reply:
There's a code below that stops every provider. So, as this function call is deferred, all data providers will be stopped and "unregistered" when this code executes.

Reply:
the code calls `Close()` which simply calls the `cancel()` function. there's no guarantee the providers have actually stopped.
https://github.com/onflow/flow-go/pull/6636/files#diff-5cbca3503bb00261318db4a8b8b1714447f7348a4d87e11aaf8b37b43b36e2bbR49-R52

Reply:
I got it. Thanks. We'd like to do it (change channel to callback) as part of #6642. It'll be much easier in terms of PR management